govcd/edgegateway.go

```diff
@@ -186,6 +186,121 @@ func (eGW *EdgeGateway) RemoveNATPortMapping(natType, externalIP, externalPort s
 
 }
 
+// AddDNATRule creates firewall DNAT rule and return refreshed existing NAT rules array or error
+func (eGW *EdgeGateway) AddDNATRule(networkHref, externalIP, externalPort, internalIP, internalPort, protocol, icmpSubType, description string) ([]*types.NatRule, error) {
+
+	task, err := eGW.AddNATRuleAsync(networkHref, "DNAT", externalIP, externalPort, internalIP, internalPort, protocol, icmpSubType, description)
+	if err != nil {
+		return nil, fmt.Errorf("error creating DNAT rule: %#v", err)
+	}
+	err = task.WaitTaskCompletion()
+	if err != nil {
+		return nil, fmt.Errorf("%s", combinedTaskErrorMessage(task.Task, err))
+	}
+
+	err = eGW.Refresh()
+	if err != nil {
+		return nil, fmt.Errorf("error refreshing edge gateway: %#v", err)
+	}
+
+	return eGW.EdgeGateway.Configuration.EdgeGatewayServiceConfiguration.NatService.NatRule, nil
+}
+
+// AddSNATRule creates firewall SNAT rule and return refreshed existing NAT rules array or error
+func (eGW *EdgeGateway) AddSNATRule(networkHref, externalIP, internalIP, description string) ([]*types.NatRule, error) {
+
+	task, err := eGW.AddNATRuleAsync(networkHref, "SNAT", externalIP, "any", internalIP, "any", "any", "", description)
+	if err != nil {
+		return nil, fmt.Errorf("error creating SNAT rule: %#v", err)
+	}
+	err = task.WaitTaskCompletion()
+	if err != nil {
+		return nil, fmt.Errorf("%s", combinedTaskErrorMessage(task.Task, err))
+	}
+
+	err = eGW.Refresh()
+	if err != nil {
+		return nil, fmt.Errorf("error refreshing edge gateway: %#v", err)
+	}
+
+	return eGW.EdgeGateway.Configuration.EdgeGatewayServiceConfiguration.NatService.NatRule, nil
+}
+
+// AddNATRuleAsync creates firewall NAT rule and return task or err
+func (eGW *EdgeGateway) AddNATRuleAsync(networkHref, natType, externalIP, externalPort, internalIP, internalPort, protocol, icmpSubType, description string) (Task, error) {
+	if !isValidProtocol(protocol) {
```

Comment:
Is the API not returning valid errors? Just checking as I think we left validations for API in some other places.

Reply:
not nice: 
```
errors.errorString = &errors.errorString{s:"operation error: task did not complete successfully:  [400:BAD_REQUEST] - [ add0f867-76df-41e8-98f4-44af29cadce2 ] Cannot update edge gateway \"vb_edge2\"\n - Cannot update edge gateway \"vb_edge2\"\n - Invalid input 'vava' provided for field featureConfigs.features[16].rules.natRulesDtos[0].protocol., error code 10050 - task error: [400 - BAD_REQUEST] [ add0f867-76df-41e8-98f4-44af29cadce2 ] Cannot update edge gateway \"vb_edge2\"\n - Cannot update edge gateway \"vb_edge2\"\n - Invalid input 'vava' provided for field featureConfigs.features[16].rules.natRulesDtos[0].protocol., error code 10050"} ("operation error: task did not complete successfully:  [400:BAD_REQUEST] - [ add0f867-76df-41e8-98f4-44af29cadce2 ] Cannot update edge gateway \"vb_edge2\"\n - Cannot update edge gateway \"vb_edge2\"\n - Invalid input 'vava' provided for field featureConfigs.features[16].rules.natRulesDtos[0].protocol., error code 10050 - task error: [400 - BAD_REQUEST] [ add0f867-76df-41e8-98f4-44af29cadce2 ] Cannot update edge gateway \"vb_edge2\"\n - Cannot update edge gateway \"vb_edge2\"\n - Invalid input 'vava' provided for field featureConfigs.features[16].rules.natRulesDtos[0].protocol., error code 10050")
```

Reply:
Oh. This is my nightmare tonight :) Thanks